### ISSUE TITLE
[nextjs app router] `ClerkProvider` breaks 404 routing

### ISSUE THREAD

**Issue by JacobJaffe** [clerk/javascript#2499]
### Preliminary Checks

- [X] I have reviewed the documentation: [https://clerk.com/docs](https://clerk.com/docs)
- [X] I have searched for existing issues: [https://github.com/clerk/javascript/issues](https://github.com/clerk/javascript/issues)
- [X] I have not already reached out to Clerk support via email or Discord (if you have, no need to open an issue here)
- [X] This issue is not a question, general help request, or anything other than a bug report directly related to Clerk. Please ask questions in our Discord community: [https://clerk.com/discord](https://clerk.com/discord).

### Reproduction / Replay Link

[https://github.com/JacobJaffe/clerk-nextjs-404-routing-demo/tree/main](https://github.com/JacobJaffe/clerk-nextjs-404-routing-demo/tree/main)

### Publishable key

pk_test_c2FmZS1zaGFkLTg2LmNsZXJrLmFjY291bnRzLmRldiQ

### Description

`ClerkProvider` wrapping the application breaks navigating away from 404 pages with `next/link`. Please note that all it takes is the `ClerkProvider` wrapping the app - I've excluded middleware in the minimal example, that doesn't change this (upgrading to the new middleware of 5.x has no effect on this bug().

**Steps to reproduce:**

See detailed instructions in the provided repo.
Navigating from a 404 page with a nextjs \`\` component does not work when the app is wrapped in a `ClerkProvider`

I have pared down the example repo to the smallest reproducible case of this, but this also occurs in more complex scenarios with custom 404 pages. Interestingly, it doesn't always occur when the `not-found.tsx` page isn't at the root level (e.g. for a dynamic route).

**Expected behavior:**

Navigating should always work.

**Actual behavior:**

The browser url indicates the route has changed, but the page does not change.

### Environment

```
System:
    OS: macOS 14.2.1
    CPU: (10) arm64 Apple M1 Max
    Memory: 15.88 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.17.0 - ~/.nvm/versions/node/v18.17.0/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 9.6.7 - ~/.nvm/versions/node/v18.17.0/bin/npm
  Browsers:
    Safari: 17.2.1
  npmPackages:
    @clerk/nextjs: ^4.29.1 => 4.29.1
    @types/node: ^20 => 20.10.6
    @types/react: ^18 => 18.2.46
    @types/react-dom: ^18 => 18.2.18
    autoprefixer: ^10.0.1 => 10.4.16
    eslint: ^8 => 8.56.0
    eslint-config-next: 14.0.4 => 14.0.4
    eslint-config-prettier: ^9.1.0 => 9.1.0
    next: 14.0.4 => 14.0.4
    postcss: ^8 => 8.4.32
    prettier: ^3.1.1 => 3.1.1
    react: ^18 => 18.2.0
    react-dom: ^18 => 18.2.0
    tailwindcss: ^3.4.0 => 3.4.0
    typescript: ^5 => 5.3.3
```

**Comment by JacobJaffe** [comment]
Please note that **this does not occur when running in development**. You'll need to build and start the server in order to reproduce this.

**Comment by JacobJaffe** [comment]
For the visual learners out there...
![2024-01-06 15 34 33](https://github.com/clerk/javascript/assets/25315679/d93151b6-aa7b-48de-9a38-c680a5c6222b)

**Comment by edtervit** [comment]
Having the same issue, all works fine in development but build fails when trying to deploy to vercel: 

`<html>
<body>

Collecting page data ...
--
13:12:43.941 | DOMException [InvalidCharacterError]: Invalid character
13:12:43.942 | at new DOMException (node:internal/per_context/domexception:53:5)
13:12:43.943 | at __node_internal_ (node:internal/util:520:10)
13:12:43.943 | at atob (node:buffer:1343:13)
13:12:43.943 | at k (/vercel/path0/.next/server/chunks/507.js:11:886)
13:12:43.943 | at /vercel/path0/.next/server/chunks/507.js:11:1080
13:12:43.943 | at P (/vercel/path0/.next/server/chunks/507.js:11:1129)
13:12:43.943 | at /vercel/path0/.next/server/chunks/507.js:94:933
13:12:43.943 | at 2475 (/vercel/path0/.next/server/chunks/507.js:94:1383)
13:12:43.943 | at t (/vercel/path0/.next/server/webpack-runtime.js:1:127)
13:12:43.943 | at 4040 (/vercel/path0/.next/server/app/_not-found.js:1:5582)
13:12:43.951 |  
13:12:43.953 | > Build error occurred
13:12:43.953 | Error: Failed to collect page data for /_not-found
13:12:43.953 | at /vercel/path0/node_modules/.pnpm/next@14.0.4_biqbaboplfbrettd7655fr4n2y/node_modules/next/dist/build/utils.js:1220:15
13:12:43.953 | at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
13:12:43.953 | type: 'Error'
13:12:43.953 | }
13:12:44.025 | ELIFECYCLE  Command failed with exit code 1.
13:12:44.063 | Error: Command "pnpm run build" exited with 1


</body>
</html>`

**Comment by kaksv** [comment]
I also got an error deploying on vercel Nextjs 14 but everything was perfect during development.

**Comment by JacobJaffe** [comment]
@kaksv  yes, this only seems to occur during while running in dev mode. Notably, this is reproducible locally though by building and serving locally.

**Comment by JacobJaffe** [comment]
There's a worse version of this bug that occurs too. It seems that only navigating to the root page breaks, so there's scenarios like:

1. Navigate to "/fake-page" -> 404 `non-found.tsx` renders
2. Navigate from there to "/real-page" -> `/real-page/page.tsx` renders
3. Navigate from there to "/" -> No navigate! 

This is really bad! In practice, a 404 page being buggy is *much* more acceptable than a 404 page breaking a *completely unrelated page*

**Comment by JacobJaffe** [comment]
I believe this error is related to how the redirects get are being attempted to be handled by both the Clerk middleware AND the NextJS server when a page is not found. Here's an error log from Vercel:

```
INFO: Clerk: The request to /lists is being redirected because there is no signed-in user, and the path is not included in `ignoredRoutes` or `publicRoutes`. To prevent this behavior, choose one of:

1. To make the route accessible to both signed in and signed out users, add "/lists" to the `publicRoutes` array passed to authMiddleware
2. To prevent Clerk authentication from running at all, pass `ignoredRoutes: ["/((?!api|trpc))(_next.*|.+\.[\w]+$)", "/lists"]` to authMiddleware
3. Pass a custom `afterAuth` to authMiddleware, and replace Clerk's default behavior of redirecting unless a route is included in publicRoutes

For additional information about middleware, please visit https://clerk.com/docs/nextjs/middleware
(This log only appears in development mode, or if `debug: true` is passed to authMiddleware)
```

Again, the expected behavior here would be that things wouldn't break, so these options aren't helpful / applicable.

**Comment by JacobJaffe** [comment]
Any update on this? 
This is really quite bad, and makes Clerk a huge risk to introduce on the App Router.

**Comment by nikosdouvlis** [comment]
Hello @JacobJaffe :)
Just wanted to let you know that we're actively working on this. I will get back to you with more details soon.

**Comment by panteliselef** [comment]
Hey @JacobJaffe turns out this issue can be replicated just by using nextjs. I have created [this issue ](https://github.com/vercel/next.js/issues/61956) in their repo. (I used the gif from this PR, I could replace it with on of ours if it bothers you.)

For some context, we are making use of `headers()` inside of ClerkProvider when this is mounted as a server component which causes the issue. While not ideal you can either add "use client" on top of `RootLayout` or for now create a custom wrapper with "use client" and only re export ClerkProvider from there.

**Comment by panteliselef** [comment]
Update: Seems like the vercel team prioritized the issue and [a PR was open](https://github.com/vercel/next.js/pull/62033) to address it. 

As a workaround, until the fix is out, `prefetch` in the `<Link/>` needs to explicitly be set to `true` or `false`.

**Comment by JacobJaffe** [comment]
@nikosdouvlis @panteliselef Thank you for the updates and for looking into this! Looking forward to the upstream fix, and thanks for relaying the workaround.

**Comment by panteliselef** [comment]
The fix should have landed in `next@14.1.1`

**Comment by clerk-cookie** [comment]
This issue has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
